### ISSUE TITLE
Very slow build time with v3.0.1

### ISSUE THREAD

**Issue by jklaise** [sphinx-doc/sphinx#7479]
**Describe the bug**
After upgrading to 3.0.1 (and hotfixing `m2r` bug https://github.com/sphinx-doc/sphinx/issues/7420), the build of the docs takes a very long time (~1500s compared to ~30s on 2.4.4). Most of the time is spent on `reading sources` which I generate dynamically with `sphinxcontrib.apidoc`. This extension just calls `sphinx-apidoc`, I have checked that the problem persists if I call `sphinx-apidoc` directly to generate sources first and then `make html` which results in a very slow build.

**To Reproduce**
To build our docs you also need `pandoc` which can be installed via package manager such as `conda` or via the OS one.

```
$ git clone https://github.com/SeldonIO/alibi.git
$ cd alibi
$ pip install -r requirements/requirements_ci.txt
$ pip uninstall m2r
$ pip install m2r @ git+https://github.com/crossnox/m2r@dev#egg=m2r
$ pip install -e .
$ make build_docs
```

**Expected behavior**
Docs build in a reasonable amount of time.

**Your project**
https://github.com/SeldonIO/alibi

**Screenshots**

**Environment info**
- OS: Arch Linux
- Python version: 3.7.3
- Sphinx version: 3.0.1
- Sphinx extensions:  
[                                                                                                                                                                                      
    'sphinx.ext.autodoc',                                                                                                                                                                           
    'sphinx.ext.doctest',                                                                                                                                                                           
    'sphinx.ext.intersphinx',                                                                                                                                                                       
    'sphinx.ext.todo',                                                                                                                                                                              
    'sphinx.ext.coverage',                                                                                                                                                                          
    'sphinx.ext.mathjax',                                                                                                                                                                           
    'sphinx.ext.ifconfig',                                                                                                                                                                          
    'sphinx.ext.viewcode',                                                                                                                                                                          
    'sphinx.ext.napoleon',                                                                                                                                                                          
    'sphinx_autodoc_typehints',                                                                                                                                                                     
    'sphinxcontrib.apidoc',                                                                      
    'nbsphinx',                                                                                                                                                                                     
    'nbsphinx_link',                                                                                                                     
    'm2r'                                                                                                                                                                                           
]        

**Additional context**
 - Updated `m2r` to use this hotfix to build docs in the first place: https://github.com/sphinx-doc/sphinx/issues/7420
 - Full `conf.py`: https://github.com/SeldonIO/alibi/blob/master/doc/source/conf.py

**Comment by fragmuffin** [comment]
I'm experiencing a similar slowdown on a closed-source project.

```
extensions = [
    'sphinx.ext.autodoc',
    'sphinx.ext.doctest',
    'sphinx.ext.intersphinx',
    'sphinx.ext.mathjax',
]
```

**Environment**
```
$ lsb_release -d
Description:	Ubuntu 18.04.3 LTS
$ sphinx-build --version
sphinx-build 3.0.1
$ python --version
Python 3.6.8
```

### Search for root cause

With all api docs removed except for the following:

```
.. automodule:: ucontrol.battery_ctrl
   :members:
   :undoc-members:
   :show-inheritance:
```

`reading sources` still takes a few seconds (much longer than it should)

When removing `:members:` **OR** `:undoc-members:` (or both), the slowdown is no longer there.
(removing `:show-inheritance:` makes no difference)

So, not a root cause, but at least it's a breadcrumb.

_edit_: I'm also using `autodoc_mock_imports` (and so is https://github.com/SeldonIO/alibi (see link in original bug report))

**Comment by tk0miya** [comment]
Fixed by #7504. It will be released as 3.0.2 soon (see #7505).
Thank you for reporting!

**Comment by fragmuffin** [comment]
Fix confirmed working with my project.
Thank you for the fix!; greatly appreciated :+1:
